akka-stream/src/main/scala/akka/stream/scaladsl/RestartFlow.scala

```diff
@@ -289,23 +289,25 @@ private abstract class RestartWithBackoffLogic[S <: Shape](
       /*
        * Upstream in this context is the wrapped stage.
        */
-      override def onUpstreamFailure(ex: Throwable) = {
+      override def onUpstreamFailure(ex: Throwable): Unit = {
         if (finishing || maxRestartsReached() || !settings.restartOn(ex)) {
           fail(out, ex)
         } else {
           logIt(s"Restarting stream due to failure [${restartCount + 1}]: $ex", OptionVal.Some(ex))
           scheduleRestartTimer()
         }
       }
-    })
 
-    setHandler(out, new OutHandler {
-      override def onPull() = sinkIn.pull()
-      override def onDownstreamFinish(cause: Throwable) = {
+      override def onPull(): Unit = sinkIn.pull()
+      override def onDownstreamFinish(cause: Throwable): Unit = {
         finishing = true
         sinkIn.cancel(cause)
       }
-    })
+    }
+
+    sinkIn.setHandler(handler)
+    setHandler(out, handler)
```

Comment:
I'm not sure I like this change. More clean to use separate handlers when they are used for two different things. Saving one allocation isn't worth it.